### DISCUSSION TITLE
[Feature] Validate File Uploads Before Actual Upload

### DISCUSSION THREAD

**Discussion by JuanDMeGon** [livewire/livewire#5245] [3. Feature ideas] [answered]
Hello!

I have faced a limitation where we, have a validation rule for file uploading of max 3Mb. Still, if the user uploads a 1Gb file, it gets uploaded to S3 in the tmp folder, and just after that, it throws the error.

Of course, that means an enormous cost in time for the user and unnecessary data transfer & storage usage.

Now. I have been working on an idea to solve this, but I want to discuss this with you before sending a PR.

Essentially, my idea is to use a validation within the WithFileUploads trait, which will avoid the call to upload:generatedSignedUrlForS3 (which starts the actual upload) if that validation fails.

Here is the code:

`startUpload@WithFileUploads` **BEFORE**:
```
public function startUpload($name, $fileInfo, $isMultiple)
    {
        if (FileUploadConfiguration::isUsingS3()) {
            throw_if($isMultiple, S3DoesntSupportMultipleFileUploads::class);

            $file = UploadedFile::fake()->create($fileInfo[0]['name'], $fileInfo[0]['size'] / 1024, $fileInfo[0]['type']);

            $this->emit('upload:generatedSignedUrlForS3', $name, GenerateSignedUploadUrl::forS3($file))->self();

            return;
        }

        $this->emit('upload:generatedSignedUrl', $name, GenerateSignedUploadUrl::forLocal())->self();
    }
```

`startUpload@WithFileUploads` **AFTER**:

```
if (FileUploadConfiguration::isUsingS3()) {

            throw_if($isMultiple, S3DoesntSupportMultipleFileUploads::class);

            $file = UploadedFile::fake()->create($fileInfo[0]['name'], $fileInfo[0]['size'] / 1024, $fileInfo[0]['type']);

            try {
                $validator = Validator::make(
                    [$name => $file],
                    [$name => $this->getRules()[$name]]
                );

                $validator->validate();
            } catch (ValidationException $e) {
                $this->emit('upload:errored', $name);

                throw $e;
            }

            $this->emit('upload:generatedSignedUrlForS3', $name, GenerateSignedUploadUrl::forS3($file))->self();

            return;
        }

        $this->emit('upload:generatedSignedUrl', $name, GenerateSignedUploadUrl::forLocal())->self();
    }
```

Essentially, as we already receive the file name under validation as `$name`, so we use it to build a validation based on the existing rules for that field.

Please, note this is only a draft. There are some possible improvements to make.

Let me know how you see this and your comments, so maybe I can send a PR with an improved version of this.

PS: This might be related to this discussion https://github.com/livewire/livewire/discussions/1400

**Comment by JuanDMeGon** [comment]
@calebporzio just tagging you here if we can get some attention on this.

I would happily create a PR to get this working in Livewire.

What do you think?

**Comment by ismailnakkar** [comment]
Having the same issue here. Would be great if this could be added to the features

**Comment by sammyaxe** [comment]
yeah, really needed feature indeed. Right now there is no way to prevent crazy large file from being uploaded

**Comment by fernandust** [comment]
same here. I think this is an important one, specially now with the new flux uploader 😉

**Comment by Tommy1223454** [comment]
This is exactly the kind of upload boundary that gets expensive fast: the app knows the file should be rejected, but only after transfer/storage already happened.

I’ve been working on this problem from the Node.js side with:
https://github.com/pompelmi/pompelmi

The core idea is to make inspect-before-storage decisions explicit, so “request accepted” and “file trusted” are not treated as the same thing.

**Comment by calebporzio** [comment, marked as answer]
Good call on this. Pre-upload validation is now supported and will ship in the next release. Thanks! https://github.com/livewire/livewire/pull/10206
